src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs

```diff
@@ -1337,4 +1366,172 @@ internal bool CanAddValue(HttpHeaderParser parser)
+
+        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+
+        public struct Enumerator : IEnumerator<KeyValuePair<string, HeaderStringValues>>
+        {
+            private Dictionary<HeaderDescriptor, object>.Enumerator _headerStoreEnumerator;
+            private bool _completed;
+
+            internal Enumerator(Dictionary<HeaderDescriptor, object>? headerStore)
+            {
+                if (headerStore != null)
+                {
+                    _headerStoreEnumerator = headerStore.GetEnumerator();
+                    _completed = false;
+                }
+                else
+                {
+                    _headerStoreEnumerator = default;
+                    _completed = true;
+                }
+                Current = default;
+            }
+
+            public bool MoveNext()
+            {
+                if (_completed || !_headerStoreEnumerator.MoveNext())
+                {
+                    _completed = true;
+                    return false;
+                }
+
+                KeyValuePair<HeaderDescriptor, object> nextPair = _headerStoreEnumerator.Current;
+
+                Current = new KeyValuePair<string, HeaderStringValues>(nextPair.Key.Name, ToStringValues(nextPair.Key, nextPair.Value));
+                return true;
+
+                static HeaderStringValues ToStringValues(HeaderDescriptor descriptor, object value)
+                {
+                    if (value is HttpHeaders.HeaderStoreItemInfo info)
+                    {
+                        if (info.OriginalRawValue != null)
+                        {
+                            return info.OriginalRawValue is string rawString ? new HeaderStringValues(rawString) : new HeaderStringValues((List<string>)info.OriginalRawValue);
+                        }
+
+                        return new HeaderStringValues(HttpHeaders.GetValuesAsStrings(descriptor, info));
+                    }
+
+                    Debug.Assert(value is string);
+                    return new HeaderStringValues((string)value);
+                }
+            }
+
+            public KeyValuePair<string, HeaderStringValues> Current { get; private set; }
+
+            public void Dispose()
+            {
+                _headerStoreEnumerator.Dispose();
+            }
+
+            object? IEnumerator.Current => Current;
+
+            void IEnumerator.Reset()
+            {
+                throw new NotImplementedException();
+            }
+        }
+    }
+
+    public readonly struct HeaderStringValues : IEnumerable<string>
+    {
+        private readonly IEnumerable<string>? _values;
+        private readonly string? _value;
+
+        public HeaderStringValues(string value)
+        {
+            _value = value;
+            _values = null;
+        }
+
+        public HeaderStringValues(IEnumerable<string> value)
```

Comment:
Do we need these constructors to be public in this release?

Reply:
Also, shouldn't this be `values` rather than `value`?

Reply:
They were approved [as public in API review](https://github.com/dotnet/runtime/issues/35126#issuecomment-670095983). Naming will be fixed.

Reply:
Still, I think Stephen's point is that these don't need to be public, so we may as well keep them private for now.